Code/src/main/java/nl/utwente/group10/ui/components/blocks/Block.java

```diff
@@ -31,124 +44,258 @@
+                        Expr errorExpr = (Expr) e.getHaskellObject();
+                        while (errorExpr instanceof Apply) {
+                            errorExpr = ((Apply) errorExpr).getChildren().get(0);
+                            index++;
+                        }
+                        // Get the Block in which the type error occurred and
+                        // set the error state for the mismatched input to true.
+                        getPane().getExprToFunction(errorExpr).getInput(index).setErrorState(true);
+                        // Indicate that an error occurred in the latest analyze attempt.
+                        getPane().setErrorOccurred(true);
+                    } else {
+                        // Without an expression we do not really now where the
+                        // type error occurred, so ignore it and let GHCi handle
+                        // it.
+                        informUnkownHaskellException();
+                        throw new TypeUnavailableException();
+                    }
+                } catch (HaskellException e) {
+                    // This should be impossible:
+                    // The possible exception would come from an Ident not found in the catalog,
+                    // A FunctionBlock without a valid Ident should never exist.
+                    informUnkownHaskellException();
+                    throw new TypeUnavailableException();
+                }
+                
+                // Now that the expressions are updated, propagate a visual refresh upwards.
+                this.setVisualState((int) newValue);
+                
+            }
         }
     }
-
-    /**
-     * Shortcut to call invalidateConnectionStateCascading(int state) with the newest state.
-     */
-    public void invalidateConnectionStateCascading() {
-        invalidateConnectionStateCascading(ConnectionCreationManager.getConnectionState());
-    }
-
+    
     /**
-     * @return Whether or not the state of the block confirms to the given newest state.
+     * ChangeListener that propagates the new VisualState to other Blocks
+     * used as input for this Block.
      */
-    public boolean connectionStateIsUpToDate(int state) {
-        return this.connectionState == state;
+    public void cascadeVisualState(ObservableValue<? extends Number> observable, Number oldValue, Number newValue) {
+        if (this instanceof InputBlock) {
+            InputBlock iblock = (InputBlock) this;
+            for (InputAnchor input : iblock.getActiveInputs()) {
+                if (input.isPrimaryConnected()) {
+                    input.getPrimaryOppositeAnchor().get().getBlock().setVisualState((int) newValue);
+                }
+            }
+        }
     }
+    
+    private void informUnkownHaskellException() {
+        String msg = "Whoops! An unkown error has occured. We're sorry, but can't really tell you more than this.";
```

Comment:
You could have this method (optionally) take an Exception and add a button to show the Exception's toString.